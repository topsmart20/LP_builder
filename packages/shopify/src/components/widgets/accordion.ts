import { blockToLiquid } from '../../functions/block-to-liquid';
import { component } from '../../constants/components';
import { style } from '../../functions/style';

// TODO: slickStyles, emulate the div wrapping layers by the slick lib etc
export const Accordion = component({
  name: 'Builder:Accordion',
  component: (block, renderOptions) => {
    const { options } = block.component!;
    const { items } = options;

    return `
    <div class="builder-accordion" style="${style({
      display: 'flex',
      alignItems: 'stretch',
      flexDirection: 'column',
      ...(options.grid && {
        flexDirection: 'row',
        alignItems: 'flex-start',
        flexWrap: 'wrap',
      }),
    })}">
      ${
        items
          ? items
              .map(
                (item: any, index: number) => `
                  <div
                    class="builder-accordion-title builder-accordion-title-closed"
                    style="${style({
                      display: 'flex',
                      flexDirection: 'column',
                      alignItems: 'stretch',
                      ...(options.grid && {
                        width: options.gridRowWidth,
                      }),
                    })}">
                    ${item.title
                      .map((child: any) => blockToLiquid(child, renderOptions))
                      .join('\n')}
                  </div>`
              )
              .join('\n')
          : ''
      }
    </div>
  `;
  },
});
