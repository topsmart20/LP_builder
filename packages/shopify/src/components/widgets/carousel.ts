import { blockToLiquid } from '../../functions/block-to-liquid';
import { component } from '../../constants/components';

// TODO: slickStyles, emulate the div wrapping layers by the slick lib etc
export const Carousel = component({
  name: 'Builder:Carousel',
  component: (block, renderOptions) => {
    const firstChild = block.children && block.children[0];
    if (firstChild && firstChild.repeat) {
      firstChild.repeat.collection = firstChild.repeat.collection + ' limit:1'
    }
    return `
    <div class="builder-carousel">
      ${
        block.children
          ? // TODO: limit repeat on first element to just one hm
            block.children
              .slice(0, 1)
              .map(child => blockToLiquid(child, renderOptions))
              .join('\n')
          : ''
      }
    </div>
  `;
  },
});
