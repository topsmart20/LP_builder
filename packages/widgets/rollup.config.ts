import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import sourceMaps from 'rollup-plugin-sourcemaps'
import typescript from 'rollup-plugin-typescript2'
import replace from 'rollup-plugin-replace'
import json from 'rollup-plugin-json'

const pkg = require('./package.json')

const libraryName = 'builder-widgets'

const resolvePlugin = resolve()

const options = {
  input: `src/${libraryName}.ts`,
  // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
  watch: {
    include: 'src/**'
  },
  external: ['vm2'],
  plugins: [
    typescript({ useTsconfigDeclarationDir: true }),
    replace({
      'process.env.NODE_ENV': JSON.stringify('production')
    }),
    // Allow json resolution
    json(),
    // Compile TypeScript files
    // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
    commonjs({
      namedExports: {
        'node_modules/d3-ease/dist/d3-ease.js': [
          'easeLinear',
          'easeQuad',
          'easeQuadIn',
          'easeQuadOut',
          'easeQuadInOut',
          'easeCubic',
          'easeCubicIn',
          'easeCubicOut',
          'easeCubicInOut',
          'easePoly',
          'easePolyIn',
          'easePolyOut',
          'easePolyInOut',
          'easeSin',
          'easeSinIn',
          'easeSinOut',
          'easeSinInOut',
          'easeExp',
          'easeExpIn',
          'easeExpOut',
          'easeExpInOut',
          'easeCircle',
          'easeCircleIn',
          'easeCircleOut',
          'easeCircleInOut',
          'easeBounce',
          'easeBounceIn',
          'easeBounceOut',
          'easeBounceInOut',
          'easeBack',
          'easeBackIn',
          'easeBackOut',
          'easeBackInOut',
          'easeElastic',
          'easeElasticIn',
          'easeElasticOut',
          'easeElasticInOut'
        ]
      }
    }),
    // Allow node_modules resolution, so you can use 'external' to control
    // which external modules to include in the bundle
    // https://github.com/rollup/rollup-plugin-node-resolve#usage
    resolvePlugin,

    // Resolve source maps to the original source
    sourceMaps()
  ]
}

export default [
  {
    ...options,
    output: {
      format: 'umd',
      file: 'dist/builder-widgets.umd.js',
      name: 'BuilderWidgets',
      sourcemap: true,
      amd: {
        id: '@builder.io/widgets'
      }
    }
  },
  {
    ...options,
    output: [
      { file: pkg.module, format: 'es', sourcemap: true },
      { file: pkg.main, format: 'cjs', sourcemap: true }
    ],
    // Do not resolve for es module build
    // TODO: should really do a cjs build too (probably for the default build instead of umd...)
    external: Object.keys(pkg.dependencies || {}).filter(name => !name.startsWith('lodash-es')),
    plugins: options.plugins.filter(plugin => plugin !== resolvePlugin).concat([
      resolve({
        only: [/^\.{0,2}\//, /lodash\-es/]
      })
    ])
  },
  {
    ...options,
    output: {
      format: 'iife',
      file: pkg.unpkg,
      name: 'BuilderWidgets',
      sourcemap: true
    }
  }
  // {
  //   ...options,
  //   output: {
  //     format: 'system',
  //     file: TODO,
  //     sourcemap: true
  //   }
  // }
]
