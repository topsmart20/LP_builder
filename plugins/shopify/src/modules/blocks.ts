import { merge } from 'lodash'
import { ShopifyProduct } from '../interfaces/shopify-product'
import { BuilderElement } from '@builder.io/sdk'

export const defaultStyles = () => ({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
  position: 'relative',
  flexShrink: '0',
  boxSizing: 'border-box',
  textAlign: 'center',
  marginTop: '15px'
})

function getParamNames(func: Function) {
  const STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/gm
  const ARGUMENT_NAMES = /([^\s,]+)/g
  const fnStr = func.toString().replace(STRIP_COMMENTS, '')
  let result = fnStr
    .slice(fnStr.indexOf('(') + 1, fnStr.indexOf(')'))
    .match(ARGUMENT_NAMES)
  if (result === null) result = []
  return result
}

// Simple token to precompile the source code to strinsg
// TODO: generate both tsCode and jsCode
export const _s = (fn: (...args: any[]) => any) =>
  'console.error("This function was unprocessed by the TS transformer!")'

const getFnBody = (fn: Function) =>
  fn
    .toString()
    .replace(
      /^\W*(function[^{]+\{([\s\S]*)\}|[^=]+=>[^{]*\{([\s\S]*)\}|[^=]+=>(.+))/i,
      '$2$3$4'
    )

// TODO: extract out args and assign the minified names to the original names
export const fnString = (fn: (...args: any[]) => any, jsCodeStyle = false) => {
  const params = getParamNames(fn)
  const body = getFnBody(fn)

  const list = jsCodeStyle
    ? [
        'data',
        'ref',
        'state',
        'update',
        'element',
        'Builder',
        'builder',
        'context'
      ]
    : [
        'state',
        'event',
        'block',
        'builder',
        'Device',
        'update',
        'Builder',
        'context'
      ]

  const hasMinifiedParamNames = params && params[0] && params[0] !== list[0]
  if (!hasMinifiedParamNames) {
    return body
  }
  let assignsStr = ''

  list.forEach((item, index) => {
    const param = params[index]
    if (param) {
      assignsStr += `var ${param} = ${item};\n`
    }
  })
  return assignsStr + body
}

export interface ShopifyProductState {
  productInfo?: ShopifyProduct & {
    price?: string
  }
  selectedProductVariant?: number
}

// TODO: why DeepPartial not work here?
export type JsElement = Partial<BuilderElement> & {
  layerLocked?: boolean
}

type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>
}

export const mergeEl = (...args: JsElement[]) =>
  merge({}, ...args.map(item => el(item)))

export const el = (info: JsElement, withParentMeta = false) =>
  merge(
    {} as BuilderElement,
    {
      '@type': '@builder.io/sdk:Element',
      responsiveStyles: { large: defaultStyles() },
      meta: {
        ...(withParentMeta && {
          requiresParent: {
            message: 'This block must be inside a "Product box" block',
            query: {
              'meta.kind': 'ProductBox'
            }
          }
        })
      }
    } as BuilderElement,
    info
  )
