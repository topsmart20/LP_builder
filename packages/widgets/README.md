# Builder Widgets

Adds widgets for Builder.io editing, such as carousels, tabs, accordions, etc.

## How to use it

When using the React SDK, just 
```ts
import '@builder.io/widgets'
```

Anywhere that you render a `<BuilderComponent ... />`, and now the widgets will register and be available in the editor and when rendering (including server side)

ðŸ‘‰See here for a real [working example in our next.js example repo](/examples/next-js/pages/[...slug].js)

Questions or feedback - contact us at help@builder.io for more info!

