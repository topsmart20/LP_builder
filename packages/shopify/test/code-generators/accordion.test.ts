import { Accordion } from '../../src/components/widgets/accordion';
import { el } from '../modules/helpers';
import { BuilderElement } from '@builder.io/sdk';

const defaultTitle: BuilderElement = {
  '@type': '@builder.io/sdk:Element',
  layerName: 'Accordion item title',
  responsiveStyles: {
    large: {
      marginTop: '10px',
      position: 'relative',
      display: 'flex',
      alignItems: 'stretch',
      flexDirection: 'column',
      paddingBottom: '10px',
    },
  },
  children: [
    {
      '@type': '@builder.io/sdk:Element',
      responsiveStyles: {
        large: {
          textAlign: 'left',
          display: 'flex',
          flexDirection: 'column',
        },
      },
      component: {
        name: 'Text',
        options: {
          text: 'I am an accordion title. Click me!',
        },
      },
    },
  ],
};

const getAccordionCode = (options: any): string =>
  Accordion(
    el({
      component: {
        options,
        name: 'Builder:Accordion',
      },
    }),
    {}
  );

describe('Accordion code generator', () => {
  test('values for all attributes', () => {
    const code = getAccordionCode({
      grid: true,
      gridRowWidth: '300px',
    });

    expect(code).toMatchSnapshot();
  });

  test('no values for attributes', () => {
    const code = getAccordionCode({});
    expect(code).toMatchSnapshot();
  });

  test('no attributes, with items', () => {
    const code = getAccordionCode({
      items: [
        {
          title: [defaultTitle],
        },
      ],
      grid: true,
    });

    expect(code).toMatchSnapshot();
  });
});
