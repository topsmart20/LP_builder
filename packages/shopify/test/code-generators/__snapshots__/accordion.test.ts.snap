// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion code generator no attributes, with items 1`] = `
"
    <div class=\\"builder-accordion\\" style=\\"display: flex;align-items: flex-start;flex-direction: row;flex-wrap: wrap;\\">
      
                  <div
                    class=\\"builder-accordion-title builder-accordion-title-closed\\"
                    style=\\"display: flex;flex-direction: column;align-items: stretch;\\">
                    
    <style>.builder-block.undefined {
  margin-top: 10px;
  position: relative;
  display: flex;
  align-items: stretch;
  flex-direction: column;
  padding-bottom: 10px;}</style>
    
    
    
    
      <div  builder-id=\\"undefined\\" class=\\"builder-block\\">
        
    <style>.builder-block.undefined {
  text-align: left;
  display: flex;
  flex-direction: column;}</style>
    
    
    
    
      <div  builder-id=\\"undefined\\" class=\\"builder-block\\">
        
      <style>
        .builder-text p:first-of-type, .builder-paragraph:first-of-type { margin: 0 } .builder-text > p, .builder-paragraph { color: inherit; line-height: inherit; letter-spacing: inherit; font-weight: inherit; font-size: inherit; text-align: inherit; font-family: inherit; }
      </style>
      <span class=\\"builder-text\\">
        I am an accordion title. Click me!
      </span>
    
      </div>
    
    
    
  
      </div>
    
    
    
  
                  </div>
    </div>
  "
`;

exports[`Accordion code generator no values for attributes 1`] = `
"
    <div class=\\"builder-accordion\\" style=\\"display: flex;align-items: stretch;flex-direction: column;\\">
      
    </div>
  "
`;

exports[`Accordion code generator values for all attributes 1`] = `
"
    <div class=\\"builder-accordion\\" style=\\"display: flex;align-items: flex-start;flex-direction: row;flex-wrap: wrap;\\">
      
    </div>
  "
`;
